exporter/exporterhelper/queued_retry.go

```diff
@@ -435,7 +435,7 @@ func (rs *retrySender) send(req internal.Request) error {
 		case <-req.Context().Done():
 			return fmt.Errorf("Request is cancelled or timed out %w", err)
 		case <-rs.stopCh:
-			return fmt.Errorf("interrupted due to shutdown %w", err)
+			return internal.NewRetryInterruptedError(fmt.Errorf("interrupted due to shutdown %w", err))
```

Comment:
If we try to requeue the data, there isn't guaranteed that the data is processed since the storage can be shutdown and the loop exits already. So the data can be requeued but not written as currentlyDispatched items in persistent storage, which still causes data loss. 

Another way we can do this is add some kinds of flushing in the persistent_queue and persistent_storage during `Stop` and make sure these queued data are processed(we can add them as currentlyDispatched items during flush). 

Do you think this way would be better?

Reply:
I'm not sure I follow. When you stop the queue, it [waits](https://github.com/open-telemetry/opentelemetry-collector/blob/cde44b428672a228ef7ee72e24762ada21a0d66c/exporter/exporterhelper/internal/persistent_queue.go#L92) for all consumers to exit before stopping the storage, so we should always be able to requeue successfully. Also, when you say the loop exits, which loop are you talking about? Putting items in the queue is synchronous, it doesn't depend on any loops or channels.

Reply:
The loop i was referring was the [loop](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/internal/persistent_storage.go#L185) in persistentStorage that process next items. 

So double checking the code again, i think i might understood it differently earlier. Do you mean this?
- consumer stop would [wait](https://github.com/open-telemetry/opentelemetry-collector/blob/cde44b428672a228ef7ee72e24762ada21a0d66c/exporter/exporterhelper/internal/persistent_queue.go#L92) for all consumer to finish.
- for each consumer right before it shuts down, if we return temporary error, they would be requeued and [produced](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/internal/persistent_queue.go#L83) in persistent storage.
- in persistent storage, two scenarios can happen:
  - the write index can be incremented so new items can be read after restart
  - or the loop can call `[getNextItem](https://github.com/open-telemetry/opentelemetry-collector/blob/cde44b428672a228ef7ee72e24762ada21a0d66c/exporter/exporterhelper/internal/persistent_storage.go#L185)` and item will be part of `currentlyDispatchedItems`
- Then after restart, the persistent storage is able to process these items

If I understand this correctly it seems cleaner than adding a new request level error.

Reply:
Yes, that's my understanding of it as well. The loop in the persistent storage only really exists so consumers can wait on a channel. Technically, an item that was removed from the storage and put in the channel by the loop counts as `dispatched`, and will be treated the same as any other dispatched items at storage start: it's going to be requeued.

Reply:
Thanks for the suggestion. I updated the PR and it looks much cleaner.